qemu/tests/block_hotplug.py

```diff
@@ -100,109 +113,70 @@ def unplug_device(vm, get_disk_cmd, device):
-            device = qdevices.QDevice(pci_type)
+            image_name = img_list[num + 1]
+            image_params = params.object_params(image_name)
+            drive_format = image_params.get("drive_format")
             if params.get("need_plug") == "yes":
                 disks_before_plug = find_disk(vm, get_disk_cmd)
-
-                if params.get("need_controller", "no") == "yes":
-                    controller_model = params.get("controller_model")
-                    controller = qdevices.QDevice(controller_model, params={"id":
-                                                  "hotadded_scsi%s" % num})
-                    bus_extra_param = params.get("bus_extra_params_%s" % img_list[num + 1])
-                    if bus_extra_param:
-                        for item in bus_extra_param.split():
-                            key, value = item.split("=", 1)
-                            qdevice_params = {key: value}
-                            controller.params.update(qdevice_params)
-                    controller.hotplug(vm.monitor)
-                    ver_out = controller.verify_hotplug("", vm.monitor)
-                    if not ver_out:
-                        err = "%s is not in qtree after hotplug" % controller_model
-                        test.fail(err)
-                    else:
-                        controller_list.append(controller)
-
-                drive = qdevices.QRHDrive("block%d" % num)
-                drive.set_param("file", find_image(img_list[num + 1]))
-                drive.set_param("format", img_format_type)
-                drive_id = drive.get_param("id")
-                drive.hotplug(vm.monitor)
-
-                device.set_param("drive", drive_id)
-                device.set_param("id", "block%d" % num)
-                if params.get("need_controller", "no") == "yes" and bool(random.randrange(2)):
-                    device.set_param("bus", controller.get_param("id")+'.0')
-                blk_extra_param = params.get("blk_extra_params_%s" % img_list[num + 1])
-                if blk_extra_param:
-                    for item in blk_extra_param.split():
-                        key, value = item.split("=", 1)
-                        device.set_param(key, value)
-                device.hotplug(vm.monitor)
-                ver_out = device.verify_hotplug("", vm.monitor)
-                if not ver_out:
-                    err = "%s is not in qtree after hotplug" % pci_type
-                    test.fail(err)
-                plug_status = utils_misc.wait_for(lambda: len(get_new_disk(disks_before_plug,
-                                                  find_disk(vm, get_disk_cmd))) != 0, pause)
+                devs = vm.devices.images_define_by_params(image_name,
+                                                          image_params, 'disk')
+                for dev in devs:
+                    dev.hotplug(vm.monitor)
+                    if isinstance(dev, qdevices.QDevice):
```

Comment:
I don't really understand why do you replace the `simple_hotplug`. It should work well and even provides some level of checking automatically while returning the same output in case it can't verify anything. Anyway even without the `simple_hotplug` it looks odd to me to call hotplug and then hardcode `""` in `verify_hotplug` on line 126.

Reply:
I mean was there anything wrong with the way I suggested the `simple_hotplug`? Was it failing for you? I think you can simply copy&paste that one and combine it with the stuff you made here.

Reply:
ok, will update it.